### PR TITLE
Allow build XContent directly from Writable

### PR DESCRIPTION
**Pull request by dnhatn**
Today, writing a Writable value to XContent in Base64 format performs these steps: (1) create a BytesStreamOutput, (2) write Writable to that output, (3) encode a copy of bytes from that output stream, (4) create a string from the encoded bytes, (5) write the encoded string to XContent. These steps allocate/use memory 5 times than writing the encode chars directly to the output of XContent.

This API would help reduce memory usage when storing a large response of an async search.

Relates #67594

### CODE REVIEW

File: libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java [hunk 1]
```diff
@@ -1047,6 +1049,33 @@ public XContentBuilder copyCurrentStructure(XContentParser parser) throws IOExce
...
+    @FunctionalInterface
+    public interface WritableValue {
+        void writeTo(OutputStream os) throws IOException;
+    }
+
+    /**
+     * Write the serialization of a {@link WritableValue} via {@link WritableValue#writeTo(OutputStream)} as a string encoded in
+     * Base64 format. This API can be used to generate XContent directly without the intermediate results to reduce memory usage.
+     */
+    public XContentBuilder writableFieldAsBase64(String name, WritableValue value) throws IOException {
```
Comment (dnhatn) [inline libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java:1064]: We can't use Writable here as it's defined in the main module.

File: libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java [hunk 2]
```diff
@@ -1047,6 +1049,33 @@ public XContentBuilder copyCurrentStructure(XContentParser parser) throws IOExce
...
+        generator.writeFieldDirectly(name, os -> {
+            os.write('\"');
+            try (OutputStream encoded = Base64.getEncoder().wrap(new FilterOutputStream(os) {
+                @Override
+                public void close() {
```
Comment (nik9000) [inline libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java:1069]: I wonder if it's be clearer not to use the try catch and have a comment about how you intentionally army closing. Not sure!

File: libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java [hunk 3]
```diff
@@ -1047,6 +1049,33 @@ public XContentBuilder copyCurrentStructure(XContentParser parser) throws IOExce
         return this;
     }
 
+    /**
+     * Used in {@link #writableFieldAsBase64(String, WritableValue)}
+     */
+    @FunctionalInterface
+    public interface WritableValue {
```
Comment (nik9000) [inline libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java:1056]: Do we have Checked consumer here? I'm not sure the new interface adds any clarity. But if you think it does I'm certaily ok with it.

File: libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java [hunk 4]
```diff
@@ -1047,6 +1049,33 @@ public XContentBuilder copyCurrentStructure(XContentParser parser) throws IOExce
...
+        generator.writeFieldDirectly(name, os -> {
+            os.write('\"');
+            try (OutputStream encoded = Base64.getEncoder().wrap(new FilterOutputStream(os) {
+                @Override
+                public void close() {
+                    // do not close the output
+                }
+            })) {
+                value.writeTo(encoded);
+            }
```
Comment (nik9000) [inline libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java:1074]: Do these quotes presume you are writing to json? I imagine there is an content type neutral way to do this but I don't know it off hand.

**Review by nik9000 — COMMENTED** [review]
Neat!

I think it's worth double checking this works outside of json. Or asserting that it is only for json somewhere.

File: libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java (hunk 2 above)
Comment (dnhatn) [reply libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java:1069]: I added a comment. We need to close the output stream that is wrapped the Base64 encoder to flush the outstanding buffer but we can't close the underlying output stream of the XContentBuilder.

File: libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java (hunk 3 above)
Comment (dnhatn) [reply libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java:1056]: We have it here, but I prefer not to expose the output stream to the caller.

**Comment by dnhatn** [comment]
> I think it's worth double checking this works outside of json. Or asserting that it is only for json somewhere.

@nik9000 Thanks so much for a quick review. I added a note in the Javadocs and a check in method writableFieldAsBase64 in https://github.com/elastic/elasticsearch/pull/73804/commits/a1a8aae89c369926e9bfb6c2db934055ceb147fa. I will look into supporting other types later as we need to solve the async search issue.

File: libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java (hunk 3 above)
Comment (dnhatn) [reply libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java]: I changed my mind. I pushed https://github.com/elastic/elasticsearch/pull/73804/commits/1affa1973ddb2743cab85b7fedaa5f0ed126135c to remove the new interface and replace it with a CheckedConsumer.

File: libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java (hunk 2 above)
Comment (nik9000) [reply libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java:1067]: Thanks! That makes perfect sense.

**Review by nik9000 — APPROVED** [review]
Thanks!

Makes sense to me.

**Comment by dnhatn** [comment]
Thanks @nik9000 :)

**Pull request merged**